### PR TITLE
Mouse wheel: vertical scrolling

### PR DESCRIPTION
**Pull request by axalix**
http://sfgui.sfml-dev.de/forum/topic155-mouse-wheel-scrolling-text-selection-and-closable-windows.html

Scrolling widgets with the mouse wheel.  Vertical scrolling only.

### CODE REVIEW

File: include/SFGUI/Widget.hpp [hunk 1]
```diff
@@ -343,6 +344,13 @@ class SFGUI_API Widget : public Object, public std::enable_shared_from_this<Widg
 		 * @param y Mouse Y position.
 		 */
 		virtual void HandleMouseMoveEvent( int x, int y );
+		
+		/** Handle mouse wheel move event.
+		 * @param delta Number of ticks the wheel has moved (positive is up, negative is down).
+		 * @param x Mouse X position.
+		 * @param y Mouse  Y position.
```
Comment (binary1248) [inline include/SFGUI/Widget.hpp]: Extraneous space.

File: src/SFGUI/Scrollbar.cpp [hunk 2]
```diff
@@ -92,6 +92,26 @@ sf::Vector2f Scrollbar::CalculateRequisition() {
 	return sf::Vector2f( mimimum_slider_length, mimimum_slider_length );
 }
 
+void Scrollbar::HandleMouseWheelEvent(int delta, int x, int y) {
+	if (GetOrientation() == Orientation::VERTICAL) {
+		auto slider_rect = GetSliderRect();
```
Comment (binary1248) [inline src/SFGUI/Scrollbar.cpp]: We don't like to declare locals if they are only used once. Move the calls into the line where they are used.

File: src/SFGUI/Scrollbar.cpp [hunk 3]
```diff
@@ -92,6 +92,26 @@ sf::Vector2f Scrollbar::CalculateRequisition() {
...
+		auto alloc = GetAllocation();
```
Comment (binary1248) [inline src/SFGUI/Scrollbar.cpp]: Same here.

File: src/SFGUI/Scrollbar.cpp [hunk 4]
```diff
@@ -92,6 +92,26 @@ sf::Vector2f Scrollbar::CalculateRequisition() {
...
+		sf::FloatRect rect(0, 0, alloc.left + slider_rect.width, alloc.height);
```
Comment (binary1248) [inline src/SFGUI/Scrollbar.cpp]: We like to initialize floating point values with floating point literals (.f suffix).

File: src/SFGUI/Scrollbar.cpp [hunk 5]
```diff
@@ -92,6 +92,26 @@ sf::Vector2f Scrollbar::CalculateRequisition() {
 	return sf::Vector2f( mimimum_slider_length, mimimum_slider_length );
 }
 
+void Scrollbar::HandleMouseWheelEvent(int delta, int x, int y) {
+	if (GetOrientation() == Orientation::VERTICAL) {
```
Comment (binary1248) [inline src/SFGUI/Scrollbar.cpp]: Missing spaces.

File: src/SFGUI/Scrollbar.cpp [hunk 6]
```diff
@@ -92,6 +92,26 @@ sf::Vector2f Scrollbar::CalculateRequisition() {
 	return sf::Vector2f( mimimum_slider_length, mimimum_slider_length );
 }
 
+void Scrollbar::HandleMouseWheelEvent(int delta, int x, int y) {
```
Comment (binary1248) [inline src/SFGUI/Scrollbar.cpp]: Missing spaces.

File: src/SFGUI/Scrollbar.cpp [hunk 7]
```diff
@@ -92,6 +92,26 @@ sf::Vector2f Scrollbar::CalculateRequisition() {
...
+	if (GetOrientation() == Orientation::VERTICAL) {
+		auto slider_rect = GetSliderRect();
+		auto alloc = GetAllocation();
+		sf::FloatRect rect(0, 0, alloc.left + slider_rect.width, alloc.height);
+	
+		if( rect.contains( static_cast<float>( x ), static_cast<float>( y ) )) {
```
Comment (binary1248) [inline src/SFGUI/Scrollbar.cpp]: Missing space at the end.

File: src/SFGUI/Scrollbar.cpp [hunk 8]
```diff
@@ -92,6 +92,26 @@ sf::Vector2f Scrollbar::CalculateRequisition() {
...
+			if( delta < 0 ) {
+				for (int i = 0; i < 5; ++i)
```
Comment (binary1248) [inline src/SFGUI/Scrollbar.cpp]: We like to use auto for indices.

File: src/SFGUI/Scrollbar.cpp [hunk 9]
```diff
@@ -92,6 +92,26 @@ sf::Vector2f Scrollbar::CalculateRequisition() {
...
+					GetAdjustment()->Increment();
+				return;
+			} else {
+				for (int i = 0; i < 5; ++i)
```
Comment (binary1248) [inline src/SFGUI/Scrollbar.cpp]: Same as above.

**Comment by binary1248** [comment]
So, if I understood the code correctly, this modification allows for the user to scroll scrollbars using their mouse wheel when the mouse cursor is hovering over the scrollbar in question?

Also, because SFGUI likes to stay up-to-date and use cutting-edge SFML features, it would be preferable if you used the floating point wheel events SFML has started to provide recently.

**Comment by axalix** [comment]
Thank you for your suggestions,  binary1248.

The suggested changes will make it possible for users to use mouse for vertical scrolling when the pointer is hovered over the scrollbar OR the the corresponding widget. Two examples can demonstrate the change: "Canvas" and "ScrolledWindowViewport" (you will need to add multiple buttons make the scrollbar visible).

Regarding, integers / float, I decided to keep "int" for now to keep the current code consistent (mostly due to insufficient time to change it everywhere, including other mouse-related methods). Although I keep my hope in others to change it...

**Comment by TankOs** [comment]
Nice patch, looks very neat. Some things that are missing:

* Horizontal scrolling when there's no vertical scrollbar (e.g. in SFGUI-Test, the lower ScrolledWindow).
* Scrolling in drop downs.
* Scrolling in sliders.

**Comment by axalix** [comment]
Can I expect these changes will be merged into master any time soon?

**Comment by binary1248** [comment]
There are still a few outstanding issues:
* When I mentioned using the newer float-based wheel events from SFML, I didn't imply changing other mouse event handling as well. SFML still provides that data using ints, so they don't need to be touched. I checked and with your patch there are a total of 2 spots in the codebase where wheel events are handled. It would be nice if you could convert them to the float variant in order for SFGUI to stay compatible with future versions of SFML.
* Like @TankOs mentioned, there are still other widgets that can benefit from the same behaviour that you implemented for containers. We don't like splitting the same kind of enhancement into separate commits for each widget type. Horizontal scrolling support can be left for a later time.
* We like to squash multiple commits that encompass a single enhancement into a single commit before merging. That way it is easier to track changes and trace history if necessary in the future.

**Comment by axalix** [comment]
I cancel my changes. Thank you everyone.

**Pull request closed without merging**